src/data-version/data-version-middleware.ts

```diff
@@ -0,0 +1,40 @@
+import { HookNextFunction, Model, Schema } from 'mongoose';
+import { InnerModelType } from '../types';
+import { getNextDataVersion } from './data-version-manager';
+
+export const addDataVersionMiddleware = (schema: Schema) => {
+    ['findOneAndUpdate', 'update', 'updateOne', 'save'].forEach(middleware => {
+        schema.pre(middleware, getDataVersionHandler());
+    });
+
+    ['updateMany', 'insertMany'].forEach(middleware => {
+        schema.pre(middleware, getDataVersionManyHandler());
+    });
+};
+
+const getDataVersionHandler = () => {
+    return async function dataVersionSetter(
+        this: InnerModelType<any>,
+        next: () => HookNextFunction,
+    ) {
+        this.dataVersion = await getNextDataVersion();
```

Comment:
data version should be updated for every request with data manipulation and not for every save operation